ddtrace/internal/ci_visibility/coverage.py

```diff
@@ -0,0 +1,126 @@
+import contextlib
+from itertools import count
+from itertools import groupby
+import json
+import os
+from typing import Iterable
+from typing import List
+from typing import Tuple
+
+from ddtrace import config
+from ddtrace.internal import compat
+from ddtrace.internal.logger import get_logger
+
+from .constants import COVERAGE_TAG_NAME
+
+
+log = get_logger(__name__)
+
+try:
+    from coverage import Coverage
+    from coverage import version_info as coverage_version
+    from coverage.numbits import numbits_to_nums
+
+    # this public attribute became private after coverage==6.3
+    EXECUTE_ATTR = "_execute" if coverage_version > (6, 3) else "execute"
+except ImportError:
+    Coverage = None  # type: ignore[misc,assignment]
+    EXECUTE_ATTR = ""
+
+
+def enabled():
+    if config._ci_visibility_code_coverage_enabled:
+        if compat.PY2:
+            return False
+        if Coverage is None:
+            log.warning(
+                "CI Visibility code coverage tracking is enabled, but the `coverage` package is not installed. "
+                "To use code coverage tracking, please install `coverage` from https://pypi.org/project/coverage/"
+            )
+            return False
+        return True
+    return False
+
+
+@contextlib.contextmanager
+def cover(span, root=None, **kwargs):
+    """Calculates code coverage on the given span and saves it as a tag"""
+    coverage_kwargs = {
+        "data_file": None,
+        "source": [root] if root else None,
+        "config_file": False,
+    }
+    coverage_kwargs.update(kwargs)
+    cov = Coverage(**coverage_kwargs)
+    cov.start()
+    cov.switch_context(str(span.trace_id))
+    yield cov
+    cov.stop()
+    span.set_tag(COVERAGE_TAG_NAME, build_payload(cov, test_id=str(span.trace_id), root=root))
```

Comment:
is setting a span tag here a mechanism to be able to read it and encode it later? If that's the case, we should remove the span tag before sending it to the backend